frontend/javascripts/admin/dataset/composition_wizard/02_upload_files.tsx

```diff
@@ -0,0 +1,192 @@
+          ):
+        </p>
+        <Upload.Dragger
+          name="files"
+          fileList={fileList}
+          onChange={handleChange}
+          beforeUpload={() => false}
+          maxCount={2}
+          multiple
+        >
+          <p className="ant-upload-drag-icon">
+            <FileExcelOutlined
+              style={{
+                margin: 0,
+                fontSize: 35,
+              }}
+            />
+          </p>
+          <p className="ant-upload-text">Drag your landmark file(s) to this area</p>
+        </Upload.Dragger>
+      </div>
+
+      <Button style={{ marginTop: 16 }} onClick={onPrev}>
+        Back
+      </Button>
+
+      <AsyncButton type="primary" style={{ marginTop: 16, marginLeft: 8 }} onClick={onNext}>
+        Next
+      </AsyncButton>
+    </div>
+  );
+}
+
+async function parseBigWarpFile(fileList: FileList): Promise<Partial<WizardContext>> {
+  const sourcePoints: Vector3[] = [];
+  const targetPoints: Vector3[] = [];
+  if (fileList.length !== 1 || fileList[0]?.originFileObj == null) {
+    throw new SoftError("Expected exactly one CSV file.");
+  }
+
+  const csv = await readFileAsText(fileList[0]?.originFileObj);
+  const lines = csv.split("\n");
+  for (const line of lines) {
+    const fields = line.split(",");
+    if (fields.length !== EXPECTED_VALUE_COUNT_PER_CSV_LINE) {
+      if (line.trim() !== "") {
+        throw new SoftError(
+          `Cannot interpret line in CSV file. Expected ${EXPECTED_VALUE_COUNT_PER_CSV_LINE} values, got ${fields.length}.`,
+        );
+      }
+      continue;
+    }
+    const [_pointName, enabled, x1, y1, z1, x2, y2, z2] = fields;
+
+    if (enabled) {
+      const source = [x1, y1, z1].map((el) => parseInt(el.replaceAll('"', ""))) as Vector3;
+      const target = [x2, y2, z2].map((el) => parseInt(el.replaceAll('"', ""))) as Vector3;
+      sourcePoints.push(source);
+      targetPoints.push(target);
+    }
+  }
+
+  return {
+    sourcePoints,
+    targetPoints,
+    datasets: [],
+    currentWizardStep: "SelectDatasets",
+  };
+}
+
+async function parseNmlFiles(fileList: FileList): Promise<Partial<WizardContext> | null> {
+  const sourcePoints: Vector3[] = [];
+  const targetPoints: Vector3[] = [];
+  if (fileList.length !== 2) {
+    throw new SoftError("Expected exactly two NML files.");
+  }
+
+  const nmlString1 = await readFileAsText(fileList[0]?.originFileObj!);
+  const nmlString2 = await readFileAsText(fileList[1]?.originFileObj!);
+
+  if (nmlString1 === "" || nmlString2 === "") {
+    throw new SoftError("NML files should not be empty.");
+  }
+
+  const { trees: trees1, datasetName: datasetName1 } = await parseNml(nmlString1);
+  const { trees: trees2, datasetName: datasetName2 } = await parseNml(nmlString2);
+
+  if (!datasetName1 || !datasetName2) {
+    throw new SoftError("Could not extract dataset names.");
+  }
+
+  const nodes1 = Array.from(
+    values(trees1)
+      .map((tree) => Array.from(tree.nodes.values())[0])
```

Comment:
This reads like only the first node of each tree is used. It would be user-friendly to show a warning once a tree with more than one node is encountered.

Reply:
I changed this so that all nodes are taken into account and that it is asserted that all tree pairs have the same amount nodes (this is done by checking that no pair is null during the `zip` iteration). I think, this is way more intuitive and flexible for users.